plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase       #-}
+-- | Optimization passes for merging independent let-bindings together.
+module Language.PlutusIR.Optimizer.Merge (mergeLets) where
+
+import           Language.PlutusIR
+import qualified Language.PlutusIR.Analysis.Dependencies as Deps
+
+import qualified Language.PlutusCore                     as PLC
+import qualified Language.PlutusCore.Name                as PLC
+
+import           Control.Lens
+import           Control.Monad.Reader
+
+import           Data.List
+import           Data.Maybe
+import qualified Data.Set                                as Set
+
+import qualified Algebra.Graph                           as G
+import qualified Algebra.Graph.ToGraph                   as T
+
+mergeLets
+    :: (PLC.HasUnique (name a) PLC.TermUnique, PLC.HasUnique (tyname a) PLC.TypeUnique)
+    => Term tyname name a
+    -> Term tyname name a
+mergeLets t =
+    let
+        depGraph :: G.Graph Deps.Node
+        depGraph = Deps.runTermDeps t
+    in
+        runReader (mergeTerm t) depGraph
+
+{- Note [Merging lets]
+We can merge independent, non-recursive, let bindings together into a single let binding, which is
+nice since it reduces the nesting of the program.
+
+We can *nearly* do this by just looking at the dependency graph and constructing level sets. However,
+nodes in the same level set may not be mergeable if their let bindings are not adjacent.
+
+So as it stands we need to traverse the term manually, performing local merging.
+-}
+mergeTerm
+    :: (MonadReader (G.Graph Deps.Node) m, PLC.HasUnique (name a) PLC.TermUnique, PLC.HasUnique (tyname a) PLC.TypeUnique)
+    => Term tyname name a
+    -> m (Term tyname name a)
+mergeTerm = \case
+    Let x NonRec bs t -> do
+        t' <- mergeTerm t
+
+        deps <- ask
+        let dependents :: Set.Set PLC.Unique
+            dependents = Set.fromList $ do
+                b <- bs
+                u <- bindingUniques b
+                {-
+                Look for the nodes which depend on the bindings defined in this let.
+
+                Logically this should be the *reachable* set. However, given that we're already topologically
+                ordered, the only way there can be a transitive dependency between two adjacent nodes is if
+                there is a direct dependency. To see this, suppose that there were a transitive dependency
+                A -> B -> C, but A and C are adjacent. This is impossible, since B must come between A and C
+                in a topological ordering.
+                -}
+                mapMaybe Deps.nodeUnique $ Set.toList $ T.preSet (Deps.Variable u) deps
+
+        let (mergeableBindings, inner) = extractMergeable dependents t'
```

Comment:
Do I understand it right that

```
let x = 1
in let y = x
   in ...
```

does not get merged to

```
let x = 1
    y = x
in ...
```

?

Reply:
Correct - that would require it to be a recursive let binding, which is more expensive to compute, so we don't do that.

Perhaps you're expecting let bindings to also be scoped top-to-bottom? I opted not to do that - if you want mutual scoping between bindings you need a recursive binding.

Reply:
> Perhaps you're expecting let bindings to also be scoped top-to-bottom?

Yep.

> I opted not to do that

I'm curious why.

Reply:
Pros:
- Order-independence. 
    - You can always reorder bindings without changing the meaning.
- Precedent. 
    - At least: Ocaml, Nix.
- Simplicity. 
    - We already have a way of doing nested scopes with nested let bindings, no need to add it another way.

Cons:
- Can't merge as many let bindings. 
    - On the other hand, we already get blocked the moment we see a recursive let.
- More annoying to write/read.
    - It's still an IR so I don't care as much about that.

If I were writing a user-facing language I'd probably do something else, but for PIR I think this way is better.

Reply:
Thanks. I there a note in the code about that? If not, I think it's worth adding these pros and cons.

Reply:
Perhaps the original design doc is the best place for this.